File: Sources/App/Classes/Preferences/TPCPathInfo.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "BuildConfig.h"

#import "TDCAlert.h"
#import "TLOLocalization.h"
#import "TPCPreferencesLocal.h"
#import "TPCPreferencesUserDefaults.h"
#import "TPCPathInfoPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@implementation TPCPathInfo

#pragma mark -
#pragma mark Utilities

+ (void)_createDirectoryAtPath:(NSString *)directoryPath
{
	NSParameterAssert(directoryPath != nil);

	if ([RZFileManager() fileExistsAtPath:directoryPath]) {
		return;
	}

	NSError *createDirectoryError = nil;

	if ([RZFileManager() createDirectoryAtPath:directoryPath
				   withIntermediateDirectories:YES
									attributes:nil
										 error:&createDirectoryError] == NO) {
		LogToConsoleError("Failed to create directory at path: '%{public}@' - %{public}@",
						  directoryPath.standardizedTildePath,
						  createDirectoryError.localizedDescription);
	}
}

+ (void)_createDirectoryAtURL:(NSURL *)directoryURL
{
	NSParameterAssert(directoryURL != nil);

	if ([RZFileManager() fileExistsAtURL:directoryURL]) {
		return;
	}

	NSError *createDirectoryError = nil;

	if ([RZFileManager() createDirectoryAtURL:directoryURL
				  withIntermediateDirectories:YES
								   attributes:nil
										error:&createDirectoryError] == NO) {
		LogToConsoleError("Failed to create directory at path: '%{public}@' - %{public}@",
						  directoryURL.standardizedTildePath,
						  createDirectoryError.localizedDescription);
	}
}

#pragma mark -
#pragma mark Application Specific

+ (NSString *)applicationBundle
{
	return RZMainBundle().bundlePath;
}

+ (NSURL *)applicationBundleURL
{
	return RZMainBundle().bundleURL;
}

+ (NSString *)applicationResources
{
	return RZMainBundle().resourcePath;
}

+ (NSURL *)applicationResourcesURL
{
	return RZMainBundle().resourceURL;
}

+ (nullable NSString *)applicationCaches
{
	NSArray *pathArray = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);

	if (pathArray.count == 0) {
		return nil;
	}

	NSString *endPath = [NSString stringWithFormat:@"/%@/", TXBundleBuildProductIdentifier];

	NSString *basePath = [pathArray.firstObject stringByAppendingPathComponent:endPath];

	[self _createDirectoryAtPath:basePath];

	return basePath;
}

+ (nullable NSURL *)applicationCachesURL
{
	NSString *sourcePath = self.applicationCaches;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

+ (nullable NSString *)groupContainer
{
	NSURL *sourceURL = self.groupContainerURL;

	if (sourceURL == nil) {
		return nil;
	}

	return sourceURL.path;
}

+ (nullable NSURL *)groupContainerURL
{
	NSURL *baseURL =
		[RZFileManager() containerURLForSecurityApplicationGroupIdentifier:TXBundleBuildGroupContainerIdentifier];

	return baseURL;
}

+ (nullable NSString *)groupContainerApplicationCaches
{
	NSURL *sourceURL = self.groupContainerApplicationCachesURL;

	if (sourceURL == nil) {
		return nil;
	}

	return sourceURL.path;
}

+ (nullable NSURL *)groupContainerApplicationCachesURL
{
	NSURL *sourceURL = self.groupContainerURL;

	if (sourceURL == nil) {
		return nil;
	}

	NSURL *baseURL = [sourceURL URLByAppendingPathComponent:@"/Library/Caches/"];

	[self _createDirectoryAtURL:baseURL];

	return baseURL;
}

+ (nullable NSString *)applicationSupport
{
	NSArray *pathArray = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);

	if (pathArray.count == 0) {
		return nil;
	}

	NSString *basePath = [pathArray.firstObject stringByAppendingPathComponent:@"/Glasstual/"];

	[self _createDirectoryAtPath:basePath];

	return basePath;
}

+ (nullable NSURL *)applicationSupportURL
{
	NSString *sourcePath = self.applicationSupport;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

+ (nullable NSString *)groupContainerApplicationSupport
{
	NSURL *sourceURL = self.groupContainerApplicationSupportURL;

	if (sourceURL == nil) {
		return nil;
	}

	return sourceURL.path;
}

+ (nullable NSURL *)groupContainerApplicationSupportURL
{
	NSURL *sourceURL = self.groupContainerURL;

	if (sourceURL == nil) {
		return nil;
	}

	NSURL *baseURL = [sourceURL URLByAppendingPathComponent:@"/Library/Application Support/Glasstual/"];

	[self _createDirectoryAtURL:baseURL];

	return baseURL;
}

+ (nullable NSString *)applicationLogs
{
	NSArray *pathArray = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);

	if (pathArray.count == 0) {
		return nil;
	}

	NSString *endPath = [NSString stringWithFormat:@"/Logs/%@/", TXBundleBuildProductIdentifier];

	NSString *basePath = [pathArray.firstObject stringByAppendingPathComponent:endPath];

	[self _createDirectoryAtPath:basePath];

	return basePath;
}

+ (nullable NSURL *)applicationLogsURL
{
	NSString *sourcePath = self.applicationLogs;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

+ (NSString *)applicationTemporary
{
	NSString *sourcePath = NSTemporaryDirectory();

	NSString *endPath = [NSString stringWithFormat:@"/%@/", TXBundleBuildProductIdentifier];

	NSString *basePath = [sourcePath stringByAppendingPathComponent:endPath];

	[self _createDirectoryAtPath:basePath];

	return basePath;
}

+ (NSURL *)applicationTemporaryURL
{
	return [NSURL fileURLWithPath:self.applicationTemporary isDirectory:YES];
}

+ (NSString *)applicationTemporaryProcessSpecific
{
	NSString *sourcePath = self.applicationTemporary;

	int processIdentifier = [[NSProcessInfo processInfo] processIdentifier];

	NSString *endPath = [NSString stringWithFormat:@"/tmp-%i", processIdentifier];

	NSString *basePath = [sourcePath stringByAppendingPathComponent:endPath];

	[self _createDirectoryAtPath:basePath];

	return basePath;
}

+ (NSURL *)applicationTemporaryProcessSpecificURL
{
	return [NSURL fileURLWithPath:self.applicationTemporaryProcessSpecific isDirectory:YES];
}

+ (NSString *)bundledExtensions
{
	return self.bundledExtensionsURL.path;
}

+ (NSURL *)bundledExtensionsURL
{
	NSURL *baseURL = self.applicationResourcesURL;

	return [baseURL URLByAppendingPathComponent:@"/Bundled Extensions/"];
}

+ (NSString *)bundledScripts
{
	return self.bundledScriptsURL.path;
}

+ (NSURL *)bundledScriptsURL
{
	NSURL *baseURL = self.applicationResourcesURL;

	return [baseURL URLByAppendingPathComponent:@"/Bundled Scripts/"];
}

+ (NSString *)bundledThemes
{
	return self.bundledThemesURL.path;
}

+ (NSURL *)bundledThemesURL
{
	NSURL *baseURL = self.applicationResourcesURL;

	return [baseURL URLByAppendingPathComponent:@"/Bundled Styles/"];
}

+ (nullable NSString *)customExtensions
{
	NSURL *sourceURL = self.customExtensionsURL;

	if (sourceURL == nil) {
		return nil;
	}

	return sourceURL.path;
}

+ (nullable NSURL *)customExtensionsURL
{
	NSURL *sourceURL = self.groupContainerApplicationSupportURL;

	if (sourceURL == nil) {
		return nil;
	}

	NSURL *baseURL = [sourceURL URLByAppendingPathComponent:@"/Extensions/"];

	[self _createDirectoryAtURL:baseURL];

	return baseURL;
}

+ (nullable NSString *)customScripts
{
	NSArray *pathArray = NSSearchPathForDirectoriesInDomains(NSApplicationScriptsDirectory, NSUserDomainMask, YES);

	if (pathArray.count == 0) {
		return nil;
	}

	NSString *basePath = pathArray.firstObject;

#if GLASSTUAL_BUILT_INSIDE_SANDBOX == 0
	[self _createDirectoryAtPath:basePath];
#endif

	return basePath;
}

+ (nullable NSURL *)customScriptsURL
{
	NSString *sourcePath = self.customScripts;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

+ (nullable NSString *)customThemes
{
	NSURL *sourceURL = self.customThemesURL;

	if (sourceURL == nil) {
		return nil;
	}

	return sourceURL.path;
}

+ (nullable NSURL *)customThemesURL
{
	NSURL *sourceURL = self.groupContainerApplicationSupportURL;

	if (sourceURL == nil) {
		return nil;
	}

	NSURL *baseURL = [sourceURL URLByAppendingPathComponent:@"/Styles/"];

	[self _createDirectoryAtURL:baseURL];

	return baseURL;
}

#pragma mark -
#pragma mark System Specific

+ (nullable NSString *)systemApplications
{
	NSArray *searchArray = NSSearchPathForDirectoriesInDomains(NSApplicationDirectory, NSSystemDomainMask, YES);

	if (searchArray.count == 0) {
		return nil;
	}

	return searchArray.firstObject;
}

+ (nullable NSURL *)systemApplicationsURL
{
	NSString *sourcePath = self.systemApplications;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

+ (NSString *)systemDiagnosticReports
{
	return @"/Library/Logs/DiagnosticReports";
}

+ (NSURL *)systemDiagnosticReportsURL
{
	return [NSURL fileURLWithPath:self.systemDiagnosticReports isDirectory:YES];
}

#pragma mark -
#pragma mark User Specific

+ (nullable NSString *)userApplicationScripts
{
	NSURL *sourceURL = self.userApplicationScriptsURL;

	if (sourceURL == nil) {
		return nil;
	}

	return sourceURL.path;
}

+ (nullable NSURL *)userApplicationScriptsURL
{
	NSURL *sourceURL = self.customScriptsURL;

	return [sourceURL URLByDeletingLastPathComponent];
}

+ (NSString *)userDiagnosticReports
{
	return self.userDiagnosticReportsURL.path;
}

+ (NSURL *)userDiagnosticReportsURL
{
	NSURL *sourceURL = self.userHomeURL;

	return [sourceURL URLByAppendingPathComponent:@"/Library/Logs/DiagnosticReports"];
}

+ (nullable NSString *)userDownloads
{
	NSArray *searchArray = NSSearchPathForDirectoriesInDomains(NSDownloadsDirectory, NSUserDomainMask, YES);

	if (searchArray.count == 0) {
		return nil;
	}

	return searchArray.firstObject;
}

+ (nullable NSURL *)userDownloadsURL
{
	NSString *sourcePath = self.userDownloads;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

+ (NSString *)userHome
{
	return [NSFileManager pathOfHomeDirectoryOutsideSandbox];
}

+ (NSURL *)userHomeURL
{
	return [NSFileManager URLOfHomeDirectoryOutsideSandbox];
}

+ (nullable NSString *)userPreferences
{
	NSArray *searchArray = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);

	if (searchArray.count == 0) {
		return nil;
	}

	return [searchArray.firstObject stringByAppendingPathComponent:@"/Preferences/"];
}

+ (nullable NSURL *)userPreferencesURL
{
	NSString *sourcePath = self.userPreferences;

	if (sourcePath == nil) {
		return nil;
	}

	return [NSURL fileURLWithPath:sourcePath isDirectory:YES];
}

@end

#pragma mark -
#pragma mark Transcript URL

@implementation TPCPathInfo (TPCPathInfoTranscriptFolderExtension)

static NSURL *_Nullable _transcriptFolderURL = nil;

+ (nullable NSString *)transcriptFolder
{
	return _transcriptFolderURL.path;
}

+ (nullable NSURL *)transcriptFolderURL
{
	return _transcriptFolderURL;
}

+ (void)setTranscriptFolderURL:(nullable NSData *)transcriptFolderURL
{
	if (_transcriptFolderURL) {
		[_transcriptFolderURL stopAccessingSecurityScopedResource];
		_transcriptFolderURL = nil;
	}

	[RZUserDefaults() setObject:transcriptFolderURL forKey:@"LogTranscriptDestinationSecurityBookmark_5"];

	[self startUsingTranscriptFolderURL];
}

+ (void)warnUserAboutStaleTranscriptFolderURL
{
	/* If logging isn't enabled, then we don't inform user of stale
	 bookmarks because they probably aren't interested in that fact. */
	if ([TPCPreferences logToDisk] == NO) {
		return;
	}

	[TDCAlert alertWithMessage:TXTLS(@"Prompts[atn-1c]")
						 title:TXTLS(@"Prompts[b7o-v4]")
				 defaultButton:TXTLS(@"Prompts[c7s-dq]")
			   alternateButton:nil];
}

+ (void)startUsingTranscriptFolderURL
{
	/* Even if user has logging disabled, we still start using the bookmark
	 so that we can present it in the Preferences window while the pop up
	 for selecting a new location is disabled. */
	NSData *bookmark = [RZUserDefaults() dataForKey:@"LogTranscriptDestinationSecurityBookmark_5"];

	if (bookmark == nil) {
		return;
	}

	BOOL resolvedBookmarkIsStale = YES;

	NSError *resolvedBookmarkError = nil;

	NSURL *resolvedBookmark = [NSURL URLByResolvingBookmarkData:bookmark
														options:NSURLBookmarkResolutionWithSecurityScope
												  relativeToURL:nil
											bookmarkDataIsStale:&resolvedBookmarkIsStale
														  error:&resolvedBookmarkError];

	if (resolvedBookmarkIsStale) {
		/* "On return, if YES, the bookmark data is stale. 
		 Your app should create a new bookmark using the 
		 returned URL and use it in place of any stored 
		 copies of the existing bookmark." */
		NSData *newBookmark = [resolvedBookmark bookmarkDataWithOptions:NSURLBookmarkCreationWithSecurityScope
										 includingResourceValuesForKeys:nil
														  relativeToURL:nil
																  error:NULL];

		if (newBookmark) {
			[self setTranscriptFolderURL:newBookmark];
		} else {
			[self warnUserAboutStaleTranscriptFolderURL];
		}

		return;
	}

	if (resolvedBookmark == nil) {
		LogToConsoleError("Error creating bookmark for URL: %{public}@", resolvedBookmarkError.localizedDescription);

		[self warnUserAboutStaleTranscriptFolderURL];

		return;
	}

	_transcriptFolderURL = resolvedBookmark;

	if ([_transcriptFolderURL startAccessingSecurityScopedResource] == NO) {
		LogToConsoleError("Failed to access bookmark");
	}
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/ApplicationSupport/TPCPathInfo.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit
import CocoaExtensions
import os

@objc(TPCPathInfo)
public final class PathInfo: NSObject {
	private static let logger = Logger(
		subsystem: Bundle.main.bundleIdentifier ?? "Glasstual",
		category: "PathInfo"
	)

	private static let transcriptBookmarkDefaultsKey = "LogTranscriptDestinationSecurityBookmark_5"
	private static let transcriptLock = NSLock()
	private nonisolated(unsafe) static var storedTranscriptFolderURL: URL?

	private static var fileManager: FileManager {
		.default
	}

	private static var productIdentifier: String {
		ApplicationInfo.applicationBundleIdentifier()
	}

	// MARK: - Directory creation

	@objc(_createDirectoryAtPath:)
	public static func createDirectory(atPath directoryPath: String) {
		createDirectory(at: URL(fileURLWithPath: directoryPath, isDirectory: true))
	}

	@objc(_createDirectoryAtURL:)
	public static func createDirectory(at directoryURL: URL) {
		if fileManager.fileExists(at: directoryURL) {
			return
		}

		do {
			try fileManager.createDirectory(at: directoryURL, withIntermediateDirectories: true)
		} catch {
			logger.error(
				"Failed to create directory at path: '\(displayPath(for: directoryURL), privacy: .public)' - \(error.localizedDescription, privacy: .public)"
			)
		}
	}

	// MARK: - Application Specific

	@objc public static var applicationBundle: String {
		Bundle.main.bundlePath
	}

	@objc public static var applicationBundleURL: URL {
		Bundle.main.bundleURL
	}

	@objc public static var applicationResources: String {
		Bundle.main.resourcePath ?? ""
	}

	@objc public static var applicationResourcesURL: URL {
		Bundle.main.resourceURL ?? Bundle.main.bundleURL
	}

	@objc public static var applicationCaches: String? {
		guard
			var basePath = firstSearchPath(
				for: .cachesDirectory,
				appending: "/\(productIdentifier)/"
			)
		else {
			return nil
		}

		basePath = applyUIReviewDirectory(toPath: basePath)
		createDirectory(atPath: basePath)

		return basePath
	}

	@objc public static var applicationCachesURL: URL? {
		fileURL(forPath: applicationCaches)
	}

	@objc public static var groupContainer: String? {
		groupContainerURL?.path
	}

	@objc public static var groupContainerURL: URL? {
		guard
			var baseURL = fileManager.containerURL(
				forSecurityApplicationGroupIdentifier: ApplicationGroup.identifier
			)
		else {
			return nil
		}

		#if DEBUG
			if let reviewDirectory = uiReviewDirectoryName {
				baseURL =
					baseURL
						.appendingPathComponent("UI Reviews", isDirectory: true)
						.appendingPathComponent(reviewDirectory, isDirectory: true)
				createDirectory(at: baseURL)
			}
		#endif

		return baseURL
	}

	@objc public static var groupContainerApplicationCaches: String? {
		groupContainerApplicationCachesURL?.path
	}

	@objc public static var groupContainerApplicationCachesURL: URL? {
		guard let sourceURL = groupContainerURL else {
			return nil
		}

		let baseURL = sourceURL.appendingPathComponent("/Library/Caches/")
		createDirectory(at: baseURL)

		return baseURL
	}

	@objc public static var applicationSupport: String? {
		guard
			var basePath = firstSearchPath(
				for: .applicationSupportDirectory,
				appending: "/Glasstual/"
			)
		else {
			return nil
		}

		basePath = applyUIReviewDirectory(toPath: basePath)
		createDirectory(atPath: basePath)

		return basePath
	}

	@objc public static var applicationSupportURL: URL? {
		fileURL(forPath: applicationSupport)
	}

	@objc public static var groupContainerApplicationSupport: String? {
		groupContainerApplicationSupportURL?.path
	}

	@objc public static var groupContainerApplicationSupportURL: URL? {
		guard let sourceURL = groupContainerURL else {
			return nil
		}

		let baseURL = sourceURL.appendingPathComponent("/Library/Application Support/Glasstual/")
		createDirectory(at: baseURL)

		return baseURL
	}

	@objc public static var applicationLogs: String? {
		guard
			var basePath = firstSearchPath(
				for: .libraryDirectory,
				appending: "/Logs/\(productIdentifier)/"
			)
		else {
			return nil
		}

		basePath = applyUIReviewDirectory(toPath: basePath)
		createDirectory(atPath: basePath)

		return basePath
	}

	@objc public static var applicationLogsURL: URL? {
		fileURL(forPath: applicationLogs)
	}

	@objc public static var applicationTemporary: String {
		let basePath = (NSTemporaryDirectory() as NSString)
			.appendingPathComponent("/\(productIdentifier)/")

		createDirectory(atPath: basePath)

		return basePath
	}

	@objc public static var applicationTemporaryURL: URL {
		URL(fileURLWithPath: applicationTemporary, isDirectory: true)
	}

	@objc public static var applicationTemporaryProcessSpecific: String {
		let processIdentifier = ProcessInfo.processInfo.processIdentifier
		let basePath = (applicationTemporary as NSString)
			.appendingPathComponent("/tmp-\(processIdentifier)")

		createDirectory(atPath: basePath)

		return basePath
	}

	@objc public static var applicationTemporaryProcessSpecificURL: URL {
		URL(fileURLWithPath: applicationTemporaryProcessSpecific, isDirectory: true)
	}

	@objc public static var bundledExtensions: String {
		bundledExtensionsURL.path
	}

	@objc public static var bundledExtensionsURL: URL {
		applicationResourcesURL.appendingPathComponent("/Bundled Extensions/")
	}

	@objc public static var bundledScripts: String {
		bundledScriptsURL.path
	}

	@objc public static var bundledScriptsURL: URL {
		applicationResourcesURL.appendingPathComponent("/Bundled Scripts/")
	}

	@objc public static var bundledThemes: String {
		bundledThemesURL.path
	}

	@objc public static var bundledThemesURL: URL {
		applicationResourcesURL.appendingPathComponent("/Bundled Styles/")
	}

	@objc public static var customExtensions: String? {
		customExtensionsURL?.path
	}

	@objc public static var customExtensionsURL: URL? {
		guard let sourceURL = groupContainerApplicationSupportURL else {
			return nil
		}

		let baseURL = sourceURL.appendingPathComponent("/Extensions/")
		createDirectory(at: baseURL)

		return baseURL
	}

	@objc public static var customScripts: String? {
		#if DEBUG
			if ProcessInfo.processInfo.environment["GLASSTUAL_UI_REVIEW_DIRECTORY"] != nil,
			   let supportURL = groupContainerApplicationSupportURL
			{
				let baseURL = supportURL.appendingPathComponent("Scripts", isDirectory: true)
				createDirectory(at: baseURL)

				return baseURL.path
			}
		#endif

		return firstSearchPath(for: .applicationScriptsDirectory)
	}

	@objc public static var customScriptsURL: URL? {
		fileURL(forPath: customScripts)
	}

	@objc public static var customThemes: String? {
		customThemesURL?.path
	}

	@objc public static var customThemesURL: URL? {
		guard let sourceURL = groupContainerApplicationSupportURL else {
			return nil
		}

		let baseURL = sourceURL.appendingPathComponent("/Styles/")
		createDirectory(at: baseURL)

		return baseURL
	}

	// MARK: - System Specific

	@objc public static var systemApplications: String? {
		firstSearchPath(for: .applicationDirectory, in: .systemDomainMask)
	}

	@objc public static var systemApplicationsURL: URL? {
		fileURL(forPath: systemApplications)
	}

	@objc public static var systemDiagnosticReports: String {
		"/Library/Logs/DiagnosticReports"
	}

	@objc public static var systemDiagnosticReportsURL: URL {
		URL(fileURLWithPath: systemDiagnosticReports, isDirectory: true)
	}

	// MARK: - User Specific

	@objc public static var userApplicationScripts: String? {
		userApplicationScriptsURL?.path
	}

	@objc public static var userApplicationScriptsURL: URL? {
		customScriptsURL?.deletingLastPathComponent()
	}

	@objc public static var userDiagnosticReports: String {
		userDiagnosticReportsURL.path
	}

	@objc public static var userDiagnosticReportsURL: URL {
		userHomeURL.appendingPathComponent("/Library/Logs/DiagnosticReports")
	}

	@objc public static var userDownloads: String? {
		firstSearchPath(for: .downloadsDirectory)
	}

	@objc public static var userDownloadsURL: URL? {
		fileURL(forPath: userDownloads)
	}

	@objc public static var userHome: String {
		FileManager.pathOfHomeDirectoryOutsideSandbox
	}

	@objc public static var userHomeURL: URL {
		FileManager.URLOfHomeDirectoryOutsideSandbox
	}

	@objc public static var userPreferences: String? {
		firstSearchPath(for: .libraryDirectory, appending: "/Preferences/")
	}

	@objc public static var userPreferencesURL: URL? {
		fileURL(forPath: userPreferences)
	}

	// MARK: - Transcript folder

	@objc public static var transcriptFolder: String? {
		transcriptFolderURL?.path
	}

	@objc public static var transcriptFolderURL: URL? {
		transcriptLock.lock()
		defer { transcriptLock.unlock() }

		return storedTranscriptFolderURL
	}

	@objc public static func setTranscriptFolderURL(_ transcriptFolderURL: Data?) {
		transcriptLock.lock()
		if let storedTranscriptFolderURL {
			storedTranscriptFolderURL.stopAccessingSecurityScopedResource()
			self.storedTranscriptFolderURL = nil
		}
		transcriptLock.unlock()

		TextualUserDefaults.shared().set(transcriptFolderURL, forKey: transcriptBookmarkDefaultsKey)
		startUsingTranscriptFolderURL()
	}

	@objc public static func startUsingTranscriptFolderURL() {
		guard let bookmark = TextualUserDefaults.shared().data(forKey: transcriptBookmarkDefaultsKey) else {
			return
		}

		var resolvedBookmarkIsStale = true
		let resolvedBookmark: URL

		do {
			resolvedBookmark = try URL(
				resolvingBookmarkData: bookmark,
				options: .withSecurityScope,
				relativeTo: nil,
				bookmarkDataIsStale: &resolvedBookmarkIsStale
			)
		} catch {
			logger.error("Error resolving bookmark for URL: \(error.localizedDescription, privacy: .public)")
			warnUserAboutStaleTranscriptFolderURL()

			return
		}

		if resolvedBookmarkIsStale {
			var newBookmark: Data?

			if resolvedBookmark.startAccessingSecurityScopedResource() {
				newBookmark = try? resolvedBookmark.bookmarkData(
					options: .withSecurityScope,
					includingResourceValuesForKeys: nil,
					relativeTo: nil
				)
				resolvedBookmark.stopAccessingSecurityScopedResource()
			}

			if let newBookmark {
				setTranscriptFolderURL(newBookmark)
			} else {
				warnUserAboutStaleTranscriptFolderURL()
			}

			return
		}

		transcriptLock.lock()
		storedTranscriptFolderURL = resolvedBookmark
		transcriptLock.unlock()

		if resolvedBookmark.startAccessingSecurityScopedResource() == false {
			logger.error("Failed to access bookmark")
		}
	}

	// MARK: - Helpers

	private static func warnUserAboutStaleTranscriptFolderURL() {
		guard TextualPreferences.logToDisk() else {
			return
		}

		_ = TDCAlert.alert(
			withMessage: PromptStrings.Logging.staleLocationBody,
			title: PromptStrings.Logging.staleLocationTitle,
			defaultButton: PromptStrings.Action.confirmation,
			alternateButton: nil
		)
	}

	private static func firstSearchPath(
		for directory: FileManager.SearchPathDirectory,
		in domainMask: FileManager.SearchPathDomainMask = .userDomainMask,
		appending suffix: String? = nil
	) -> String? {
		guard let firstURL = FileManager.default.urls(for: directory, in: domainMask).first else {
			return nil
		}
		let firstPath = firstURL.path

		guard let suffix else {
			return firstPath
		}

		return (firstPath as NSString).appendingPathComponent(suffix)
	}

	private static func fileURL(forPath path: String?) -> URL? {
		guard let path else {
			return nil
		}

		return URL(fileURLWithPath: path, isDirectory: true)
	}

	private static func displayPath(for url: URL) -> String {
		(url as NSURL).textualStandardizedTildePath ?? url.path
	}

	private static func applyUIReviewDirectory(toPath basePath: String) -> String {
		#if DEBUG
			guard let reviewDirectory = uiReviewDirectoryName else {
				return basePath
			}

			return ((basePath as NSString).appendingPathComponent("UI Reviews") as NSString)
				.appendingPathComponent(reviewDirectory)
		#else
			return basePath
		#endif
	}

	#if DEBUG
		private static var uiReviewDirectoryName: String? {
			let reviewDirectory = ProcessInfo.processInfo.environment["GLASSTUAL_UI_REVIEW_DIRECTORY"] ?? ""

			guard reviewDirectory.isEmpty == false else {
				return nil
			}

			return (reviewDirectory as NSString).lastPathComponent
		}
	#endif
}
